### PR TITLE
DM-996 part 2

### PR DESCRIPTION
**Pull request by SkyToGround**
### Description of work

* Build nodes now use GCC8.
* Changed `namespace Module` into `namespace WriterModule`.
* The class `Module::WriterBase` is now instead `WriterModule::Base`.
* I have split the files  _HDFWriterModule.*_  into _WriterModuleBase.*_ and _WriterRegistrar.*_.
* Added support for (and requirement of) module names into the writer registration code.
* Added/changed unit tests for new writer module registration code.

### Nominate for Group Code Review

- [ ] Nominate for code review

### CODE REVIEW

File: src/WriterRegistrar.cpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                                     return CItem.second.Id == FlatbufferID;
+                                   });
+  if (FoundFactory == Factories.end()) {
+    throw std::out_of_range("Unable to find factory with the id: " +
+                            FlatbufferID);
+  }
+  return (*FoundFactory).second.FactoryPtr;
+}
+
+ModuleFactory const &find(std::string const &FlatbufferID,
```
Comment (mattclarke) [inline src/WriterRegistrar.cpp]: This and the one below only seem to be used in tests.
Is this prep for future work?

File: src/WriterRegistrar.cpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  }
+  FoundItem = std::find_if(std::cbegin(Factories), std::cend(Factories),
+                           [&ModuleName](auto const &CItem) {
+                             return CItem.second.Name == ModuleName;
+                           });
+  if (FoundItem != Factories.end()) {
+    if (not FlatbufferID.empty() and FoundItem->second.Id != FlatbufferID) {
+      throw std::runtime_error(
+          "The provided flatbuffer id (" + FlatbufferID +
+          ") does not correspont to (" + FoundItem->second.Id +
```
Comment (mattclarke) [inline src/WriterRegistrar.cpp]:
```suggestion
          ") does not correspond to (" + FoundItem->second.Id +
```

File: src/WriterRegistrar.cpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  if (FoundFactory == Factories.end()) {
+    throw std::out_of_range("Unable to find factory with the id: " +
+                            FlatbufferID);
+  }
+  return (*FoundFactory).second.FactoryPtr;
+}
+
+ModuleFactory const &find(std::string const &FlatbufferID,
+                          std::string const &ModuleName) {
+  auto const &Factories = getFactories();
```
Comment (mattclarke) [inline src/WriterRegistrar.cpp]: Check `if (not FlatbufferID.empty())` at the top as a guard condition.
It is empty then we aren't going to get anything useful, right?

File: src/WriterRegistrar.cpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+ModuleFactory const &find(IdNameHash ModuleHash) {
+  return getFactories().at(ModuleHash).FactoryPtr;
+}
+
+void clear() { getFactories().clear(); }
+
+// cppcheck-suppress unusedFunction
+void addWriterModule(std::string const &FlatbufferID,
+                     std::string const &ModuleName, ModuleFactory Value) {
+  auto &Factories = getFactories();
```
Comment (mattclarke) [inline src/WriterRegistrar.cpp]: Move this down to where it is used. I.e. between line 96 and 97

File: src/WriterRegistrar.cpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  if (FlatbufferID.size() != 4) {
+    throw std::runtime_error(
+        "The number of characters in the Flatbuffer id string must be 4.");
+  }
+  auto ModuleHash = getWriterModuleHash(FlatbufferID, ModuleName);
+  if (Factories.find(ModuleHash) != Factories.end()) {
```
Comment (mattclarke) [inline src/WriterRegistrar.cpp]: If we used emplace then can all this code be replaced like we saw yesterday?
Something like `
  Factories.emplace(ModuleHash, FactoryInfo{Value, FlatbufferID, ModuleName});` might be sufficient

File: src/WriterRegistrar.h [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+void addWriterModule(std::string const &FlatbufferID,
+                     std::string const &ModuleName, ModuleFactory Value);
+
+/// \brief Get `ModuleFactory for a given flatbuffer id.
+///
+/// \return Matching `ModuleFactory`.
+ModuleFactory const &find(std::string const &FlatbufferID);
+
+/// \brief Get module factory for a given flatbuffer id and/or module name.
+/// \param[in] FlatbufferID A four character flatbuffer id. Will be ignored if
```
Comment (mattclarke) [inline src/WriterRegistrar.h]: Can we lose the [in]s please? It is unnecessary noise in most cases.
A new line after the \brief makes my rapidly failing eyes happy.

Generally, I think we should aim to keep the doxygen as simple as possible then there is less chance of it being out of date.

Obviously it is a different story if we are supplying a public API...

File: src/kafka-to-nexus.cpp [hunk 7]
```diff
@@ -63,18 +64,15 @@ int main(int argc, char **argv) {
...
     for (auto &ReaderPair :
          FileWriter::FlatbufferReaderRegistry::getReaders()) {
       fmt::print("---- {}\n", ReaderPair.first);
     }
-    fmt::print("\n--Identifiers of HDFWriterModule factories\n");
-    for (auto &WriterPair :
-         FileWriter::HDFWriterModuleRegistry::getFactories()) {
-      fmt::print("---- {}\n", WriterPair.first);
+    fmt::print("\n--Known writer modules\n");
+    for (auto &WriterPair : WriterModule::Registry::getFactoryIdsAndNames()) {
```
Comment (mattclarke) [inline src/kafka-to-nexus.cpp]:
```suggestion
    for (auto const &WriterPair : WriterModule::Registry::getFactoryIdsAndNames()) {
```

File: src/tests/writer_module/WriterRegistrationTests.cpp [hunk 8]
```diff
@@ -8,64 +8,174 @@
...
+  auto UsedHash =
+      WriterModule::Registry::getWriterModuleHash(UsedKey, UsedName);
+  {
+    WriterModule::Registry::Registrar<StubWriterModule> RegisterIt(UsedKey,
+                                                                   UsedName);
+  }
+  EXPECT_NE(WriterModule::Registry::find(UsedHash), nullptr);
+}
+
+TEST_F(WriterRegistrationTest, HashKeyNotFound) {
```
Comment (mattclarke) [inline src/tests/writer_module/WriterRegistrationTests.cpp]:
```suggestion
TEST_F(WriterRegistrationTest, HashKeyNotFoundThrows) {
```

File: src/tests/writer_module/WriterRegistrationTests.cpp [hunk 9]
```diff
@@ -8,64 +8,174 @@
...
+  auto UsedHash =
+      WriterModule::Registry::getWriterModuleHash(WrongKey, UsedName);
+  {
+    WriterModule::Registry::Registrar<StubWriterModule> RegisterIt(UsedKey,
+                                                                   UsedName);
+  }
+  EXPECT_THROW(WriterModule::Registry::find(UsedHash), std::out_of_range);
+}
+
+TEST_F(WriterRegistrationTest, FacoryIdsAndNames) {
```
Comment (mattclarke) [inline src/tests/writer_module/WriterRegistrationTests.cpp]:
```suggestion
TEST_F(WriterRegistrationTest, FactoryIdsAndNamesRegisterCorrectly) {
```

File: src/tests/writer_module/WriterRegistrationTests.cpp [hunk 10]
```diff
@@ -8,64 +8,174 @@
...
+  std::string UsedKey("t3mp");
+  std::string UsedName("some_module_name");
+  {
+    WriterModule::Registry::Registrar<StubWriterModule> RegisterIt(UsedKey,
+                                                                   UsedName);
+  }
+  EXPECT_NE(WriterModule::Registry::find("", UsedName), nullptr);
+}
+
+TEST_F(WriterRegistrationTest, FailFindModuleUsingName) {
```
Comment (mattclarke) [inline src/tests/writer_module/WriterRegistrationTests.cpp]:
```suggestion
TEST_F(WriterRegistrationTest, FindModuleThrowsForUnknownName) {
```

File: src/tests/writer_module/WriterRegistrationTests.cpp [hunk 11]
```diff
@@ -8,64 +8,174 @@
...
+  std::string UsedKey("t3mp");
+  std::string UsedName("some_module_name");
+  {
+    WriterModule::Registry::Registrar<StubWriterModule> RegisterIt(UsedKey,
+                                                                   UsedName);
+  }
+  EXPECT_NE(WriterModule::Registry::find(UsedKey, UsedName), nullptr);
+}
+
+TEST_F(WriterRegistrationTest, FailFindModuleUsingNameAndId) {
```
Comment (mattclarke) [inline src/tests/writer_module/WriterRegistrationTests.cpp]:
```suggestion
TEST_F(WriterRegistrationTest, FindModuleThrowsForUnknownNameAndUnknownId) {
```

File: src/tests/writer_module/WriterRegistrationTests.cpp [hunk 12]
```diff
@@ -8,64 +8,174 @@
...
+  std::string UsedName("some_module_name");
+  {
+    WriterModule::Registry::Registrar<StubWriterModule> RegisterIt(UsedKey,
+                                                                   UsedName);
+  }
+  EXPECT_THROW(WriterModule::Registry::find("tst2", "Some other name"),
+               std::out_of_range);
+}
+
+TEST_F(WriterRegistrationTest, FailFindModuleUsingNameAndWrongId) {
```
Comment (mattclarke) [inline src/tests/writer_module/WriterRegistrationTests.cpp]:
```suggestion
TEST_F(WriterRegistrationTest, FindModuleThrowsForUnknownId) {
```

File: src/WriterModuleBase.h [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+  /// \param HDFAttributes Additional attributes as defined in the Nexus
+  /// structure which the module should write to the file. Because the
+  /// writer module is free to create the structure and data sets according to
+  /// its needs, it must also take the responsibility to write these
+  /// attributes.
+  /// \param HDFAttributes Json string of the attributes associated with the
+  /// stream, as defined by the "attributes" key in the Nexus structure.
+  ///
+  /// \return The result.
+  virtual InitResult init_hdf(hdf5::node::Group &HDFGroup,
```
[the remark this answers is not in the document]
Comment (SkyToGround) [reply src/WriterModuleBase.h]: I agree. This of course requires us to make changes to the unit tests, writer modules and init-code. I suggest we create a separate ticket for that change.

File: src/WriterModuleBase.h [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+                              std::string const &HDFAttributes) = 0;
+
+  /// \brief Reopen the HDF objects which are used by this writer module.
+  ///
+  /// \param InitParameters Contains most importantly the \p HDFGroup into
+  /// which this module should write its data.
+  ///
+  /// \return The result.
+  virtual InitResult reopen(hdf5::node::Group &HDFGroup) = 0;
```
[the remark this answers is not in the document]
Comment (SkyToGround) [reply src/WriterModuleBase.h]: Ditto.

File: src/WriterRegistrar.cpp [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+namespace WriterModule {
+namespace Registry {
+
+struct FactoryInfo {
+  ModuleFactory FactoryPtr;
+  std::string Id;
+  std::string Name;
+};
+
+std::map<IdNameHash, FactoryInfo> &getFactories() {
```
[the remark this answers is not in the document]
Comment (SkyToGround) [reply src/WriterRegistrar.cpp]: Sure. Done.

File: src/WriterRegistrar.cpp (hunk 1 above)
Comment (SkyToGround) [reply src/WriterRegistrar.cpp]: Yes, to limit the size of the PR.

File: src/WriterRegistrar.cpp (hunk 5 above)
Comment (matthew-d-jones) [reply src/WriterRegistrar.cpp]: Also I think you can `std::move(Value)` here to avoid a copy

**Comment by matthew-d-jones** [comment]
I assume the docker image failing to build now is due to the compiler version change

File: src/WriterModuleBase.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (mattclarke) [reply src/WriterModuleBase.h]: Done

**Pull request merged**